File: ruby/spec/array_spec.rb
# frozen_string_literal: true

describe Array do
  context '空の配列を用意する' do
    Given(:sut) { [] }
    Then { sut.empty? }

    context '値を追加する' do
      When { sut << 1 << 2 << 3 }
      Then { sut.size == 3 }
      Then { sut[0] == 1 }
      Then { sut[1] == 2 }
      Then { sut[2] == 3 }

      context '要素を合計する' do
        When(:sum) do
          result = 0
          sut.each do |value|
            result += value
          end
          result
        end
        Then { sum == 6 }
      end
    end
  end

  context '要素あり配列を用意する' do
    Given(:sut) { [4, 5, 6] }
    Then { !sut.empty? }
    Then { sut.size == 3 }
    Then { sut[0] == 4 }
    Then { sut[1] == 5 }
    Then { sut[2] == 6 }
  end
end

File: typescript/src/array.spec.ts
describe("Array", () => {
    describe('空の配列を用意する', () => {
        let sut: number[] = [];

        it('配列は空である', () => {
            expect(sut.length).toEqual(0);
        });

        describe('値を追加する', () => {
            beforeAll(() => {
                sut.push(1);
                sut.push(2);
                sut.push(3);
            });

            it('要素数=3', () => {
                expect(sut.length).toEqual(3);
            });

            it('追加した順に格納されている', () => {
                expect(sut[0]).toEqual(1);
                expect(sut[1]).toEqual(2);
                expect(sut[2]).toEqual(3);
            });

            describe('要素を合計する', () => {
                let sum = 0;

                beforeAll(() => {
                    for (let value of sut) {
                        sum += value;
                    }
                });

                it('結果は6', () => {
                    expect(sum).toEqual(6);
                });
            })
        });
    });

    describe('要素あり配列を用意する', () => {
        let sut = [4, 5, 6];

        it('要素数=3', () => {
            expect(sut.length).toEqual(3);
        });
    });
});
